File: DownloadHistoryController.m
//
//  DownloadHistoryController.m
//  Get_iPlayer GUI
//
//  Created by Thomas Willson on 10/15/09.
//  Copyright 2009 __MyCompanyName__. All rights reserved.
//

#import "DownloadHistoryController.h"
#import "DownloadHistoryEntry.h"
#import "Programme.h"
#import "NSFileManager+DirectoryLocations.h"


@implementation DownloadHistoryController
- (instancetype)init
{
	if (!(self = [super init])) return nil;
	[self readHistory:self];
	[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(addToHistory:) name:@"AddProgToHistory" object:nil];
	return self;
}
- (void)readHistory:(id)sender
{
	NSLog(@"Read History");
	if ([historyArrayController.arrangedObjects count] > 0)
		[historyArrayController removeObjectsAtArrangedObjectIndexes:[NSIndexSet indexSetWithIndexesInRange:NSMakeRange(0, [historyArrayController.arrangedObjects count])]];
	
    NSString *historyFilePath = [[[NSFileManager defaultManager] applicationSupportDirectory] stringByAppendingPathComponent:@"download_history"];
	NSFileHandle *historyFile = [NSFileHandle fileHandleForReadingAtPath:historyFilePath];
	NSData *historyFileData = [historyFile readDataToEndOfFile];
	NSString *historyFileInfo = [[NSString alloc] initWithData:historyFileData encoding:NSUTF8StringEncoding];
	
	if (historyFileInfo.length > 0)
	{
		NSString *string = [NSString stringWithString:historyFileInfo];
		NSUInteger length = string.length;
		NSUInteger paraStart = 0, paraEnd = 0, contentsEnd = 0;
		NSMutableArray *array = [NSMutableArray array];
		NSRange currentRange;
		while (paraEnd < length) {
			[string getParagraphStart:&paraStart end:&paraEnd
						  contentsEnd:&contentsEnd forRange:NSMakeRange(paraEnd, 0)];
			currentRange = NSMakeRange(paraStart, contentsEnd - paraStart);
			[array addObject:[string substringWithRange:currentRange]];
		}
		for (NSString *entry in array) {
            NSArray *components = [entry componentsSeparatedByString:@"|"];
            NSString *pid = components[0];
            NSString *showName = components[1];
            NSString *episodeName = components[2];
            NSString *type = components[3];
            NSString *someNumber = components[4];
            NSString *downloadFormat = components[5];
            NSString *downloadPath = components[6];

            DownloadHistoryEntry *historyEntry = [[DownloadHistoryEntry alloc] initWithPID:pid
                                                                                  showName:showName
                                                                               episodeName:episodeName
                                                                                      type:type
                                                                                someNumber:someNumber
                                                                            downloadFormat:downloadFormat
                                                                              downloadPath:downloadPath];
            [historyArrayController addObject:historyEntry];
        }
	}
	NSLog(@"end read history");
}

- (IBAction)writeHistory:(id)sender
{
	if (!runDownloads || [sender isEqualTo:self])
	{
		NSLog(@"Write History to File");
		NSArray *currentHistory = historyArrayController.arrangedObjects;
		NSMutableString *historyString = [[NSMutableString alloc] init];
		for (DownloadHistoryEntry *entry in currentHistory)
		{
			[historyString appendFormat:@"%@\n", [entry entryString]];
		}
        NSString *historyPath = [[[NSFileManager defaultManager] applicationSupportDirectory] stringByAppendingPathComponent:@"download_history"];
		NSData *historyData = [historyString dataUsingEncoding:NSUTF8StringEncoding];
		NSFileManager *fileManager = [NSFileManager defaultManager];
		if (![fileManager fileExistsAtPath:historyPath])
        {
			if (![fileManager createFileAtPath:historyPath contents:historyData attributes:nil])
            {
                NSAlert *alert = [[NSAlert alloc] init];
                alert.informativeText = @"Please submit a bug report saying that the history file could not be created.";
                alert.messageText = @"Could not create history file!";
                [alert addButtonWithTitle:@"OK"];
                [alert runModal];
                [self addToLog:@"Could not create history file!"];
            }
        }
		else
        {
            NSError *writeToFileError;
			if (![historyData writeToFile:historyPath options:NSDataWritingAtomic error:&writeToFileError])
            {
                NSAlert *alert = [[NSAlert alloc] init];
                alert.informativeText = @"Please submit a bug report saying that the history file could not be written to.";
                alert.messageText = @"Could not write to history file!";
                [alert addButtonWithTitle:@"OK"];
                [alert runModal];
            }
        }
	}
	else
	{
        NSAlert *alert = [[NSAlert alloc] init];
        alert.informativeText = @"Your changes have been discarded.";
        alert.messageText = @"Download History cannot be edited while downloads are running.";
        [alert addButtonWithTitle:@"OK"];
        [alert runModal];
		[historyWindow close];
	}
	[saveButton setEnabled:NO];
	[historyWindow setDocumentEdited:NO];
}

-(IBAction)showHistoryWindow:(id)sender
{
	if (!runDownloads)
	{
		if (!historyWindow.documentEdited) [self readHistory:self];
		[historyWindow makeKeyAndOrderFront:self];
		saveButton.enabled = historyWindow.documentEdited;
	}
	else
	{
        NSAlert *alert = [NSAlert new];
        alert.messageText = @"Download History cannot be edited while downloads are running.";
		[alert runModal];
	}
}

-(IBAction)removeSelectedFromHistory:(id)sender;
{
	if (!runDownloads)
	{
		[saveButton setEnabled:YES];
		[historyWindow setDocumentEdited:YES];
		[historyArrayController remove:self];
	}
	else
	{
        NSAlert *alert = [NSAlert new];
        alert.messageText = @"Download History cannot be edited while downloads are running.";
		[alert runModal];
		[historyWindow close];
	}
}
- (IBAction)cancelChanges:(id)sender
{
	[historyWindow setDocumentEdited:NO];
	[saveButton setEnabled:NO];
	[historyWindow close];
}
- (void)addToHistory:(NSNotification *)note
{
	[self readHistory:self];
	NSDictionary *userInfo = note.userInfo;
	Programme *prog = [userInfo valueForKey:@"Programme"];
	DownloadHistoryEntry *entry = [[DownloadHistoryEntry alloc] initWithPID:prog.pid showName:prog.seriesName episodeName:prog.episodeName type:nil someNumber:@"251465" downloadFormat:@"flashhigh" downloadPath:@"/"];
	[historyArrayController addObject:entry];
	[self writeHistory:self];
}
- (void)addToLog:(NSString *)logMessage
{
	[[NSNotificationCenter defaultCenter] postNotificationName:@"AddToLog" object:self userInfo:@{@"message": logMessage}];
}
@end

File: DownloadHistoryController.swift
//
//  DownloadHistoryController.swift
//  Get iPlayer Automator
//
//  Created by Scott Kovatch on 7/10/25.
//

import Cocoa
import CocoaLumberjackSwift

@objc public class DownloadHistoryController: NSObject {
    @IBOutlet weak var historyArrayController: NSArrayController!
    @IBOutlet weak var historyWindow: NSWindow!
    @IBOutlet weak var saveButton: NSButton!
    @IBOutlet weak var cancelButton: NSButton!

    var runDownloads: Bool = false

    public override func awakeFromNib() {
        super.awakeFromNib()
        readHistory()
        NotificationCenter.default.addObserver(self, selector: #selector(addToHistory(_:)), name: NSNotification.Name("AddProgToHistory"), object: nil)
    }

    private var historyFilePath: String? {
        guard let dir = FileManager.default.applicationSupportDirectory() else { return nil }
        return (dir as NSString).appendingPathComponent("download_history")
    }

    func readHistory() {
        DDLogVerbose("Read History")
        guard let historyFilePath = historyFilePath else { return }

        if let existing = historyArrayController.content as? [Any], !existing.isEmpty {
            historyArrayController.remove(contentsOf: existing)
        }

        guard let historyFile = FileHandle(forReadingAtPath: historyFilePath) else { return }
        let historyFileData = historyFile.readDataToEndOfFile()
        guard let historyFileInfo = String(data: historyFileData, encoding: .utf8), !historyFileInfo.isEmpty else { return }

        let historyEntries = historyFileInfo.components(separatedBy: .newlines)
        for entry in historyEntries where !entry.isEmpty {
            let components = entry.components(separatedBy: "|")
            guard components.count >= 7 else { continue }
            let historyEntry = DownloadHistoryEntry()
            historyEntry.pid = components[0]
            historyEntry.showName = components[1]
            historyEntry.episodeName = components[2]
            historyEntry.type = components[3]
            historyEntry.someNumber = components[4]
            historyEntry.downloadFormat = components[5]
            historyEntry.downloadPath = components[6]
            historyArrayController.addObject(historyEntry)
        }
        DDLogVerbose("End read history")
    }

    @IBAction func writeHistory(_ sender: Any?) {
        if !runDownloads {
            DDLogVerbose("Write History to File")
            guard let currentHistory = historyArrayController.arrangedObjects as? [DownloadHistoryEntry] else { return }
            let historyString = currentHistory.map { $0.entryString }.joined(separator: "\n") + "\n"
            guard let historyPath = historyFilePath else {
                return
            }
            guard let historyData = historyString.data(using: .utf8) else { return }
            let fileManager = FileManager.default
            if !fileManager.fileExists(atPath: historyPath) {
                if !fileManager.createFile(atPath: historyPath, contents: historyData, attributes: nil) {
                    let alert = NSAlert()
                    alert.informativeText = "Please submit a bug report saying that the history file could not be created."
                    alert.messageText = "Could not create history file!"
                    alert.addButton(withTitle: "OK")
                    alert.runModal()
                    DDLogWarn("\(String(describing: self)): Could not create history file!")
                }
            } else {
                do {
                    try historyData.write(to: URL(fileURLWithPath: historyPath), options: .atomic)
                } catch {
                    let alert = NSAlert()
                    alert.informativeText = "Please submit a bug report saying that the history file could not be written to."
                    alert.messageText = "Could not write to history file!"
                    alert.addButton(withTitle: "OK")
                    alert.runModal()
                }
            }
        } else {
            let alert = NSAlert()
            alert.informativeText = "Your changes have been discarded."
            alert.messageText = "Download History cannot be edited while downloads are running."
            alert.addButton(withTitle: "OK")
            alert.runModal()
            historyWindow.close()
        }
        saveButton.isEnabled = false
        historyWindow.isDocumentEdited = false
    }

    @IBAction func showHistoryWindow(_ sender: Any?) {
        if !runDownloads {
            if !(historyWindow.isDocumentEdited) {
                readHistory()
            }
            historyWindow.makeKeyAndOrderFront(self)
            saveButton.isEnabled = historyWindow.isDocumentEdited
        } else {
            let alert = NSAlert()
            alert.messageText = "Download History cannot be edited while downloads are running."
            alert.runModal()
        }
    }

    @IBAction func removeSelectedFromHistory(_ sender: Any?) {
        if !runDownloads {
            saveButton.isEnabled = true
            historyWindow.isDocumentEdited = true
            historyArrayController.remove(sender)
        } else {
            let alert = NSAlert()
            alert.messageText = "Download History cannot be edited while downloads are running."
            alert.runModal()
            historyWindow.close()
        }
    }

    @IBAction func cancelChanges(_ sender: Any?) {
        historyWindow.isDocumentEdited = false
        saveButton.isEnabled = false
        historyWindow.close()
    }

    @objc func addToHistory(_ note: Notification) {
        readHistory()
        guard let userInfo = note.userInfo,
              let programs = userInfo["Programmes"] as? [Programme] else { return }
        let now = Int(Date().timeIntervalSince1970)
        for prog in programs {
            let entry = DownloadHistoryEntry()
            entry.pid = prog.pid
            entry.showName = prog.seriesName
            entry.episodeName = prog.episodeName
            entry.someNumber = "\(now)"
            entry.type = prog.typeDescription
            entry.downloadFormat = "flashhigh"
            entry.downloadPath = prog.path
            historyArrayController.addObject(entry)
        }
        writeHistory(self)
    }
}
